e2e/e2e_test.go

```diff
@@ -0,0 +1,2610 @@
+				t.Logf("Function returned 200 but response is not a CloudEvent (attempt %v, elapsed %v)", attemptCount, elapsed.Round(time.Second))
+				t.Logf("Content-Type: %s, Ce-Specversion: %s", contentType, ceSpecVersion)
+				lastWarnTime = time.Now()
+			}
+			continue
+		}
+
+		if res.StatusCode == 500 {
+			body, _ := io.ReadAll(res.Body)
+			t.Log("500 response received; canceling retries.")
+			t.Logf("Response: %s\n", body)
+			return false
+		}
+
+		elapsed := time.Since(startTime)
+		if elapsed > cfg.warnThreshold && time.Since(lastWarnTime) >= cfg.warnInterval {
+			t.Logf("Function responded with status %d (attempt %v, elapsed %v)", res.StatusCode, attemptCount, elapsed.Round(time.Second))
+			lastWarnTime = time.Now()
+		}
+	}
+
+	t.Logf("Could not validate CloudEvents function after %v (timeout %v)", time.Since(startTime).Round(time.Second), cfg.timeout)
+	return false
+}
+
+// waitForContent returns true if there is a service listening at the
+// given addresss which responds HTTP OK with the given string in its body.
+// returns false if the.
+func waitForContent(t *testing.T, address, content string, options ...waitOption) (ok bool) {
+	return waitFor(t, address, content, "expected content not found", options...)
+}
+
+// waitFor an endpoint to return an OK response which includes the given
+// content.
+//
+// If the Function returns a 500, it is considered a positive test failure
+// by the implementation and retries are discontinued.
+//
+// TODO:  Implement a --output=json flag on `func run` and update all
+// callers currently passing localhost:8080 with this calculated value.
+//
+// Reasoning: This will be a false negative if port 8080 is being used
+// by another process.  This will fail because, if a service is already running
+// on port 8080, Functions will automatically choose to run the next higher
+// unused port.  And this will be a false positive if there happens to be
+// a service not already running on the port which happens to implement an
+// echo.  For example there is another process outside the E2Es which is
+// currently executing a `func run`
+// Note that until this is implemented, this temporary implementation also
+// forces single-threaded test execution.
+func waitFor(t *testing.T, address, content, errMsg string, options ...waitOption) (ok bool) {
+	t.Helper()
+	cfg := waitCfg{
+		timeout:       2 * time.Minute,
+		interval:      5 * time.Second,
+		warnThreshold: 30 * time.Second,
+		warnInterval:  10 * time.Second,
+	}
+	for _, o := range options {
+		o(&cfg)
+	}
+
+	var (
+		mismatchLast     = ""    // cache the last content for squelching purposes.
+		mismatchReported = false // note that the given content was reported
+	)
+
+	startTime := time.Now()
+	lastWarnTime := time.Time{}
+	attemptCount := 0
+
+	for time.Since(startTime) < cfg.timeout {
+		attemptCount++
+		time.Sleep(cfg.interval)
+		// t.Logf("GET %v\n", address)
```

Comment:
Remove or uncomment.

```suggestion
          t.Logf("GET %v\n", address)
```